File: FE/src/models/UserModel.ts
// Address
export interface Address {
  street: string;
  ward: string;
  district: string;
  city: string;
}

export const defaultAddress: Address = {
  street: "",
  ward: "",
  district: "",
  city: "",
};

// Personal Info
export interface PersonalInfo {
  full_name: string;
  tel: string;
  email: string;
  date_of_birth: string;
  gender: string;
  address: Address;
}

export const defaultPersonalInfo: PersonalInfo = {
  full_name: "",
  tel: "",
  email: "",
  date_of_birth: "",
  gender: "",
  address: defaultAddress,
};

// Identification
export interface Identification {
  id_number: string;
  issue_date?: string;
  issue_place: string;
}

export const defaultIdentification: Identification = {
  id_number: "",
  issue_date: "",
  issue_place: "",
};

// Face Data
export interface FaceData {
  face_image_url?: string;
  embedding_version?: string;
  embeddings?: string[];
}

export const defaultFaceData: FaceData = {
  face_image_url: "",
  embedding_version: "",
  embeddings: [],
};

// Face Image
export interface FaceImage {
  image_url?: string;
  image_type?: string;
  uploaded_at?: string;
  file_size?: number;
  file_format?: string;
}

export const defaultFaceImage: FaceImage = {
  image_url: "",
  image_type: "",
  uploaded_at: "",
  file_size: 0,
  file_format: "",
};

// User Model
export interface UserModel {
  _id: string;
  user_id: string;
  personal_info: PersonalInfo;
  identification: Identification;
  face_data: FaceData;
  images: FaceImage[];
}

export const defaultUserModel: UserModel = {
  _id: "",
  user_id: "",
  personal_info: defaultPersonalInfo,
  identification: defaultIdentification,
  face_data: defaultFaceData,
  images: [],
};

// Utility parser functions
export const parseUserModel = (json: any): UserModel => ({
  _id: json._id ?? "",
  user_id: json.user_id ?? "",
  personal_info: {
    full_name: json.personal_info?.full_name ?? "",
    tel: json.personal_info?.tel ?? "",
    email: json.personal_info?.email ?? "",
    date_of_birth: json.personal_info?.date_of_birth ?? "",
    gender: json.personal_info?.gender ?? "",
    address: {
      street: json.personal_info?.address?.street ?? "",
      ward: json.personal_info?.address?.ward ?? "",
      district: json.personal_info?.address?.district ?? "",
      city: json.personal_info?.address?.city ?? "",
    },
  },
  identification: {
    id_number: json.identification?.id_number ?? "",
    issue_date: json.identification?.issue_date ?? "",
    issue_place: json.identification?.issue_place ?? "",
  },
  face_data: {
    face_image_url: json.face_data?.face_image_url ?? "",
    embedding_version: json.face_data?.embedding_version ?? "",
    embeddings: json.face_data?.embeddings ?? [],
  },
  images:
    json.images?.map((img: any) => ({
      image_url: img.image_url ?? "",
      image_type: img.image_type ?? "",
      uploaded_at: img.uploaded_at ?? "",
      file_size: img.file_size ?? 0,
      file_format: img.file_format ?? "",
    })) ?? [],
});

File: app_checkin/lib/models/user_model.dart
class Address {
  final String street;
  final String ward;
  final String district;
  final String city;

  Address({
    this.street = '',
    this.ward = '',
    this.district = '',
    this.city = '',
  });

  factory Address.fromJson(Map<String, dynamic> json) {
    return Address(
      street: json['street'] ?? '',
      ward: json['ward'] ?? '',
      district: json['district'] ?? '',
      city: json['city'] ?? '',
    );
  }

  Map<String, dynamic> toJson() {
    return {'street': street, 'ward': ward, 'district': district, 'city': city};
  }
}

class PersonalInfo {
  final String fullName;
  final String tel;
  final String email;
  final String dateOfBirth;
  final String gender;
  final Address address;

  PersonalInfo({
    this.fullName = '',
    this.dateOfBirth = '',
    this.gender = '',
    this.email = '',
    this.tel = '',
    Address? address,
  }) : address = address ?? Address();

  factory PersonalInfo.fromJson(Map<String, dynamic> json) {
    return PersonalInfo(
      fullName: json['full_name'] ?? '',
      tel: json['tel'] ?? '',
      email: json['email'] ?? '',
      dateOfBirth: json['date_of_birth'] ?? '',
      gender: json['gender'] ?? '',
      address: Address.fromJson(json['address'] ?? {}),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'full_name': fullName,
      'date_of_birth': dateOfBirth,
      'tel': tel,
      'email': email,
      'gender': gender,
      'address': address.toJson(),
    };
  }
}

class Identification {
  final String idNumber;
  final String? issueDate;
  final String issuePlace;

  Identification({this.idNumber = '', this.issueDate, this.issuePlace = ''});

  factory Identification.fromJson(Map<String, dynamic> json) {
    return Identification(
      idNumber: json['id_number'] ?? '',
      issueDate: json['issue_date'],
      issuePlace: json['issue_place'] ?? '',
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id_number': idNumber,
      'issue_date': issueDate,
      'issue_place': issuePlace,
    };
  }
}

class FaceData {
  final String? faceImageUrl;
  final String? embeddingVersion;
  final List<String>? embeddings;

  FaceData({this.embeddings, this.faceImageUrl, this.embeddingVersion});

  factory FaceData.fromJson(Map<String, dynamic> json) {
    return FaceData(
      faceImageUrl: json['face_image_url'] ?? '',
      embeddingVersion: json['embedding_version'] ?? '',
      embeddings:
          (json['embeddings'] as List<dynamic>?)
              ?.map((e) => e.toString())
              .toList() ??
          [],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'face_image_url': faceImageUrl,
      'embedding_version': embeddingVersion,
      'embeddings': embeddings,
    };
  }
}

class FaceImage {
  final String? imageUrl;
  final String? imageType;
  final String? uploadedAt;
  final int? fileSize;
  final String? fileFormat;

  FaceImage({
    this.imageUrl,
    this.imageType,
    this.uploadedAt,
    this.fileSize,
    this.fileFormat,
  });

  factory FaceImage.fromJson(Map<String, dynamic> json) {
    return FaceImage(
      imageUrl: json['image_url'] ?? '',
      imageType: json['image_type'] ?? '',
      uploadedAt: json['uploaded_at'] ?? '',
      fileSize: json['file_size'] ?? 0,
      fileFormat: json['file_format'] ?? '',
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'image_url': imageUrl,
      'image_type': imageType,
      'uploaded_at': uploadedAt,
      'file_size': fileSize,
      'file_format': fileFormat,
    };
  }
}

class UserModel {
  final String id;
  final String userId;
  final String? fcmToken;
  final PersonalInfo personalInfo;
  final Identification identification;
  final FaceData faceData;
  final List<FaceImage> images;

  UserModel({
    required this.id,
    required this.userId,
    this.fcmToken,
    PersonalInfo? personalInfo,
    Identification? identification,
    FaceData? faceData,
    List<FaceImage>? images,
  }) : personalInfo = personalInfo ?? PersonalInfo(),
       identification = identification ?? Identification(),
       faceData = faceData ?? FaceData(),
       images = images ?? [];

  factory UserModel.fromJson(Map<String, dynamic> json) {
    return UserModel(
      id: json['_id'] ?? '',
      userId: json['user_id'] ?? '',
      fcmToken: json['fcm_token'] ?? '',
      personalInfo: PersonalInfo.fromJson(json['personal_info'] ?? {}),
      identification: Identification.fromJson(json['identification'] ?? {}),
      faceData: FaceData.fromJson(json['face_data'] ?? {}),
      images: (json['images'] as List<dynamic>? ?? [])
          .map((e) => FaceImage.fromJson(e as Map<String, dynamic>))
          .toList(),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      '_id': id,
      'user_id': userId,
      'fcm_token': fcmToken,
      'personal_info': personalInfo.toJson(),
      'identification': identification.toJson(),
      'face_data': faceData.toJson(),
      'images': images.map((e) => e.toJson()).toList(),
    };
  }
}
